airflow/plugins/operators/airtable_to_warehouse.py

```diff
@@ -1,126 +1,192 @@
-import pandas as pd
-import re
+import gzip
 import os
+import pandas as pd
+import pendulum
 
 from pyairtable import Table
-from calitp import save_to_gcfs, to_snakecase, write_table
+from pydantic import BaseModel
+from slugify import slugify
+from typing import Optional, Dict
+from calitp import to_snakecase
+from calitp.storage import get_fs
 
 from airflow.models import BaseOperator
 
+AIRTABLE_API_KEY = os.environ["CALITP_AIRTABLE_API_KEY"]
 
-def airtable_to_df(
-    air_base_id,
-    air_table_name,
-    id_name="__id",
-    rename_fields=None,
-    column_prefix=None,
-    api_key=None,
-):
-    """Download an airtable as a DataFrame.
-
-    Note that airtable records have rows structured as follows:
-        [{"id", "fields": {colname: value, ...}, ...]
 
-    This function applies renames in the following order.
-
-        1. rename id
-        2. rename fields
-        3. apply column prefix (to columns not renamed by 1 or 2)
+def fix_array_containing_null(arr):
+    """
+    BigQuery doesn't allow arrays that contain null values --
+    see: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#array_nulls
+    Therefore we need to manually replace nulls with falsy values according
+    to the data in the array.
     """
+    # we already know the array has a null value before we call this
+    # check whether there are any non-null values
+    if len(set(arr)) > 1:
+        non_null_entries = set(arr) - {None}
+        non_null_types = {type(entry) for entry in non_null_entries}
+        if non_null_types in [{int}, {float}]:
+            new_array = [x if x is not None else -1 for x in arr]
+        else:
+            new_array = [x if x is not None else "" for x in arr]
+    else:
+        # if it only has null values, just return an empty array --
+        # this collapses a distinction in lookup cases
+        # between records in the current table that have a record in the foreign
+        # table where the lookup field is null in the foreign table
+        # (these would return an array with a null entry like [null])
+        # vs. records in the current table that have no corresponding record
+        # (these would return a null array like [])
+        # in the foreign table
+        # however, this distinction should be recoverable from other fields
+        new_array = []
+    return new_array
+
+
+def make_arrays_bq_safe(raw_data):
+    safe_data = {}
+    for k, v in raw_data.items():
+        if type(v) == dict:
+            make_arrays_bq_safe(v)
+        elif type(v) == list:
+            if None in v:
```

Comment:
I'd just process all non-empty lists

Reply:
I had it just process all lists because I didn't see why checking `len(v) > 0` was preferable to checking `None in v`. If you think that distinction is important I can add it back in.